LCLSGeneral/LCLSGeneral/POUs/Diagnostics/FB_EtherCATDiag.TcPOU

```diff
@@ -0,0 +1,575 @@
+	TMOUT		:= tTimeout,
+	BOXNAME		=> strName
+);
+
+IF NOT fbGetSlaveName.BUSY THEN
+	fbGetSlaveName(START:= FALSE);
+
+	(* add scanned info *)
+	arrSlaveInfoScanned[iDiagIndex[iIdx]].nIndex 	:= iDiagIndex[iIdx] + 1;
+	IF arrScannedSlaveInfo[iDiagIndex[iIdx]].nAddr <> 0 THEN
+		IF NOT fbGetSlaveName.ERR THEN
+			arrSlaveInfoScanned[iDiagIndex[iIdx]].sName := strName;
+		END_IF
+	ELSE
+		arrSlaveInfoScanned[iDiagIndex[iIdx]].sType		:= '';
+	END_IF
+
+	IF (iDiagIndex[iIdx] < nScannedSlaves) THEN
+		arrSlaveInfoScanned[iDiagIndex[iIdx]].sType		:= F_ConvProductCodeToString(arrScannedSlaveInfo[iDiagIndex[iIdx]].stSlaveIdentity);
+	ELSE
+		arrSlaveInfoScanned[iDiagIndex[iIdx]].sType		:= '';
+	END_IF
+
+	arrSlaveInfoScanned[iDiagIndex[iIdx]].nECAddr	:= arrScannedSlaveInfo[iDiagIndex[iIdx]].nAddr;
+
+	IF arrSlaveInfoScanned[iDiagIndex[iIdx]].sName <> arrSlaveInfo[iDiagIndex[iIdx]].sName THEN
+		arrSlaveInfoScanned[iDiagIndex[iIdx]].bDifferentName := TRUE;
+	ELSE
+		arrSlaveInfoScanned[iDiagIndex[iIdx]].bDifferentName := FALSE;
+	END_IF
+
+	IF arrSlaveInfoScanned[iDiagIndex[iIdx]].nECAddr <> arrSlaveInfo[iDiagIndex[iIdx]].nECAddr THEN
+		arrSlaveInfoScanned[iDiagIndex[iIdx]].bDifferentAddr := TRUE;
+	ELSE
+		arrSlaveInfoScanned[iDiagIndex[iIdx]].bDifferentAddr := FALSE;
+	END_IF
+
+	IF arrSlaveInfoScanned[iDiagIndex[iIdx]].sType <> arrSlaveInfo[iDiagIndex[iIdx]].sType THEN
+		arrSlaveInfoScanned[iDiagIndex[iIdx]].bDifferentType := TRUE;
+	ELSE
+		arrSlaveInfoScanned[iDiagIndex[iIdx]].bDifferentType := FALSE;
+	END_IF
+
+	IF iIdx < 3 THEN
+		iIdx := iIdx + 1;
+		iState := iState - 2;
+	ELSE
+		iIdx := 0;
+		iState := 10;
+
+		FOR I := 0 TO 3 DO
+			IF iDiagPort[I] <> 0 THEN
+				arrDiagSlaveInfo[I] := arrSlaveInfo[iDiagIndex[I]];
+				arrDiagSlaveInfoScanned[I] := arrSlaveInfoScanned[iDiagIndex[I]];
+			ELSE
+				MEMSET(ADR(arrDiagSlaveInfo[I]), 0, SIZEOF(arrDiagSlaveInfo[I]));
+				MEMSET(ADR(arrDiagSlaveInfoScanned[I]), 0, SIZEOF(arrDiagSlaveInfoScanned[I]));
+			END_IF
+		END_FOR
+	END_IF
+END_IF
+
+
+]]></ST>
+      </Implementation>
+    </Action>
+    <Action Name="M_GetSlaveAdresses" Id="{057b915b-50ee-41a3-b261-5584ed00ac7b}">
+      <Implementation>
+        <ST><![CDATA[fbGetSlaveAddresses(
+	sNetId		:= sMasterNetID,
+	pAddrBuf	:= ADR(arrSlaveAddresses),
+	cbBufLen	:= SIZEOF(arrSlaveAddresses),
+	bExecute	:= TRUE,
+	tTimeout	:= tTimeout,
+	nSlaves		=> iNumOfSlavesRead
+);
+
+IF NOT fbGetSlaveAddresses.bBusy THEN
+	fbGetSlaveAddresses(bExecute:= FALSE);
+	IF NOT fbGetSlaveAddresses.bError THEN
+		FOR I := 0 TO MIN(iNumOfSlavesRead, iSLAVEADDR_ARR_SIZE) DO
+			arrSlaveInfo[I].nECAddr := arrSlaveAddresses[I];
+		END_FOR
+	END_IF
+	iState := iState + 1;
+END_IF
+]]></ST>
+      </Implementation>
+    </Action>
+    <Action Name="M_GetSlaveIdentity" Id="{6436ec18-e4f4-4415-8713-025490e22c62}">
+      <Implementation>
+        <ST><![CDATA[fbGetSlaveIdentity(
+	sNetId		:= sMasterNetID,
+	nSlaveAddr	:= arrSlaveInfo[iDiagIndex[iIdx]].nECAddr,
```

Comment:
Consider replacing all of the `iDiagIndex[iIdx]` with a reference - repeating it every other line is inefficient in performance and source code verbosity, and might even be error-prone at times.

All actions should also verify `__ISVALIDREF` at the beginning (https://infosys.beckhoff.com/english.php?content=../content/1033/tc3_plc_intro/18014401038647691.html&id=).